subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java

```diff
@@ -19,30 +19,48 @@
 import net.rubygrapefruit.platform.file.FileWatcher;
 import net.rubygrapefruit.platform.file.FileWatcherCallback;
 import org.gradle.internal.vfs.watch.FileWatcherRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.Set;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static org.gradle.internal.vfs.watch.FileWatcherRegistry.Type.CREATED;
 import static org.gradle.internal.vfs.watch.FileWatcherRegistry.Type.INVALIDATE;
 import static org.gradle.internal.vfs.watch.FileWatcherRegistry.Type.MODIFIED;
 import static org.gradle.internal.vfs.watch.FileWatcherRegistry.Type.REMOVED;
 
 public class AbstractEventDrivenFileWatcherRegistry implements FileWatcherRegistry {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractEventDrivenFileWatcherRegistry.class);
+
     private final FileWatcher watcher;
     private final AtomicInteger numberOfReceivedEvents = new AtomicInteger();
     private final AtomicBoolean unknownEventEncountered = new AtomicBoolean();
+    private final AtomicReference<Throwable> errorWhileReceivingFileChanges = new AtomicReference<>();
 
     public AbstractEventDrivenFileWatcherRegistry(Set<Path> roots, FileWatcherCreator watcherCreator, ChangeHandler handler) {
         this.watcher = createWatcher(roots, watcherCreator, handler);
     }
 
     private FileWatcher createWatcher(Set<Path> roots, FileWatcherCreator watcherCreator, ChangeHandler handler) {
-        FileWatcher watcher = watcherCreator.createWatcher((type, path) -> handleEvent(type, path, handler));
+        FileWatcher watcher = watcherCreator.createWatcher(new FileWatcherCallback() {
+            @Override
+            public void pathChanged(Type type, String path) {
+                handleEvent(type, path, handler);
+            }
+
+            @Override
+            public void reportError(Throwable ex) {
+                LOGGER.error("Error while receiving file changes", ex);
+                errorWhileReceivingFileChanges.compareAndSet(null, ex);
+                handler.handleLostState();
```

Comment:
I guess this is where we should stop the thing, no? Though calling `stop()` from here would be bad, as we are on the native thread here, and we'd get into a deadlock.

Reply:
Yeah, I think stopping would make sense (or even failing the build?). Like you noted, I don't have a way to stop from right here, so what do you suggest we do?

It is also fine to keep the thing running, I think. What is the bad side effect?

Reply:
I guess it depends on what caused the problem in the first place. We might end up with logging many large exception stack traces.

Reply:
As to how to tear the thing down, I'm not sure. One way is to add a `terminateAsync()`-style call, returning a `Future` perhaps. We could make all the calls async, too.